### ISSUE TITLE
Enum-ification

### ISSUE THREAD

**Issue by whitequark** [whitequark/solvespace#123]
Give names to enums and convert all `int` parameters and fields that are actually enums to use those names.

Use the `unreachable` intrinsic and Clang warnings to make sure that if/else chains and switches always cover all enum variants.

**Comment by Evil-Spirit** [comment]
I think we should use
`enum Operation : uint8_t {`
to ensure the type size.

**Comment by Evil-Spirit** [comment]
@whitequark, Just pushed draft. I need you opinion about:
* types of enums(always uint32_t or only for saved things)
* naming
* location (inside class, outside class, using "enum class", etc)

**Comment by whitequark** [comment]
> always uint32_t or only for saved things

Yes, always uint32_t for now. This makes it easier to do bindings, because the type of enum is not specified in C or C++, and setting everything to just uint32_t removes that ambiguity. Also please don't change the numbering in this changeset.

> location (inside class, outside class, using "enum class", etc)

Just enum inside class everywhere. That should be fine in all cases I can see.

> naming

In e.g. `Group::GroupType` the second `Group` is extraneous, just call it `Group::Type`. In e.g. `ClassifyType` and `CombineType` it's better to call them `ClassifyHow` and `CombineHow`.

**Comment by Evil-Spirit** [comment]
I think ClassifyHow not sounds like typename, but like variable name
```
void f(ClassifyHow how)  //  extra how
void f(Classify how)          // like real language, but in general I think it's bad practice
void f(ClassifyType how)  // sounds nomal
```

**Comment by whitequark** [comment]
> void f(ClassifyHow how)   //  extra how

This is the same thing as "Group group" etc. The variable name is the shortening of type name. So I think that's OK.

**Comment by Evil-Spirit** [comment]
> Just enum inside class everywhere
Note what we can't do it everywhere because we have poor header/cpp splitting, so in some number of cases we have to use enum before class definition.
I don't know, is where something like: 
```
// fwd decl
enum Class::Type : uint32_t;
```

**Comment by whitequark** [comment]
Yes you can do that in C++11.

**Comment by Evil-Spirit** [comment]
> Yes you can do that in C++11.
That's good.

**Comment by whitequark** [comment]
Fixed in f33ddc94, c890d663c and e2ea8c027.
